### DISCUSSION TITLE
Operating with dictionary without checking each time key

### DISCUSSION THREAD

**Discussion by SolarOsQuiere** [dotnet/csharplang#5560] [Language Ideas]
Here its the question where i get the idea:
https://stackoverflow.com/questions/69766649/operations-over-dictionary-key-without-system-collections-generic-keynotfoundexc

Im going to explain it:
I have a simple dictionary called result:

```
Dictionary<String,String> results=new();
results["a"]="a";
results["b"]="b";
results["c"]="c";
```
To simplify the example my dictionary only contains 3 letter keys a,b,c. But sometimes it will not contain one of this values or even none, (it will always be initiallized). suppose this situation:

```
Dictionary<String,String> results=new();
if(anyconditionA) results["a"]=s;
if(anyconditionB)results["b"]="b";
if(anyconditionC)results["c"]="c";
```
So each time i want to operate with this dictionary i have to check the key value:
 var test= results["a"]; -> throws System.Collections.Generic.KeyNotFoundException
 if anycontitionA is not true. so to solve this i do:

```
if(results.ContainsKey("a"){
  someOperation(results["a"]);
}
```
Same checking using tryGetValue
so if i have many values code looks like:

if(results.ContainsKey("a"){someOperation(results["a"]);}
if(results.ContainsKey("b"){... }
if(results.ContainsKey("c"){someOperation(results["d"]);}

What about a checking like null operator in list or extension method like this:

 
```
public static bool DoIfExists<TKey, TValue>(this Dictionary<TKey, TValue> self, TKey key, Action<TValue> action)
{
     if (!self.TryGetValue(key, out var value))
     return false;
    
     action(value);
     return true;
}
```

So then i can :
`results.DoIfExists("somekey", SomeOperation);`
or 
`results["somekey"]?.SomeOperation(x=> x.Value)`

Then code looks smarter:

```
results["a"]?.SomeOperation(x=>x.key)
results["b"]?.SomeOperation(x=>x.key)
results["c"]?.SomeOperation(x=>x.key)
```

**Comment by jnm2** [comment]
`results["a"]?.SomeOperation(x => x.key)` has an existing, valid meaning in today's C#. It would not be able to change to mean something else.

(It means: call the indexer on `result` using argument `"a"`, then invoke a `SomeOperation` method on whatever the indexer returns if the indexer does not return null.)

**Comment by theunrepentantgeek** [comment]
There's a difference between a dictionary that doesn't have a particular key, and a dictionary that contains null for that same key - so your proposed syntax doesn't really solve the problem stated.

Plus, doing repeated lookups for the same key value has problems. Not only is it inefficient, but you might run into concurrency problems where the dictionary is modified between `ContainsKey("a")` and `["a"]`. _Even if concurrency is not a concern on your current project_ writing code like this is a bad habit to establish.

Instead of this code (reformatted to usual defaults):
``` csharp
if (results.ContainsKey("a"))
{
    someOperation(results["a"]);
}
```

you can instead write
``` csharp
if (results.TryGetValue("a", out var v))
{
    someOperation(v);
}
```

If you want to skip **null** values, extend the condition:
``` csharp
if (results.TryGetValue("a", out var v) && v != null )
{
    someOperation(v);
}
```

**Reply by jnm2** [reply]
The original post mentioned this TryGetValue approach and asks for something more ergonomic, the same way `?.` is more ergonomic than `if`.

**Reply by SolarOsQuiere** [reply]
Yeah exactly that an operator to implicit-checking

**Reply by SolarOsQuiere** [reply]
> lookups

Yeah but why not an operator of extension method to simplify the checking? . I used the `? `to clarify the example of what im looking for.

**Comment by RikkiGibson** [comment]
Consider using [`CollectionExtensions.GetValueOrDefault`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.collectionextensions.getvalueordefault?view=net-6.0#System_Collections_Generic_CollectionExtensions_GetValueOrDefault__2_System_Collections_Generic_IReadOnlyDictionary___0___1____0_).

(looks like this was already mentioned on stackoverflow :smile:)
